### PR TITLE
[DRAFT] Switching to equinox; overhauling the field.

### PR DESCRIPTION
**Pull request by GJBoth**
## The current situation

Currently our field is based off Flax's `struct.PyTreeNode`, and internally everything has the same 5D shape: `[batch height width wavelength scalar/vector]`. Additionally, due to issues with `jit` and `vmap`, `dx`, and `spectrum` are not part of the pytree, and they thus don't get transformed by vmap. For example

```
@jax.jit
def forward_chromatix(dx):
    u = jnp.ones((1, 512, 512, 1, 1))
    return cf.generic_field(dx, 1.0, 1.0, u, jnp.zeros_like(u))
```
throws an error. Internally this results in a lot of complicated and messy shape wrangling (separate access methods for every parameter), and externally we have lots of empty axes for end users. If you just want to simulate a scalar single wavelength (which is the majority of users), why do they need to deal with 5D tensors of which only 2 axes are non-empty?

In short, the problems are:
1. Returning a field from a vmapped / jitted function tends to throw of weird errors. 
2. A lot of empty axes.
3. Internal complexity.

TL;DR: Lots of confusion for users.

## Proposed solution

This draft PR proposes a way around this, *without* (large) breaking changes. The main ideas are:

1) Switch to [equinox](https://github.com/patrick-kidger/equinox) modules, as it makes life easier, and contains a lot of useful goodies.
2) Make more field types, which all have different shapes, depending on amount of wavelengths, scalar vs vector, etc.
3) Get rid of the batch axis, as this resulted in a lot of shape wrangling.
4) Every class has an `dim` attributes mapping names to axis, indicating where the axis is. This is a bare-bones replacement for named axes, which unfortunately doesn't seem to be happening in the jax ecosystem.
5) Minorly, this PR also adds in some jaxtyping for shape annotation. 

Instead of `ScalarField` and `VectorField` - we now have fields like `ScalarCoherentField` - shape `[h w]`, or `SpectralScalarCoherentField`, of shape `[h w l]`. Despite the changes in the shape, they are still internally consistent, so we can use the same function on these two fields, for example:

```
def phase_change(field: AbstractField , z, n=1.0) -> AbstractField:
    L_sq = field.spectrum * z / n
    phase = (jnp.pi / L_sq) * l2_sq_norm(field.grid)
    return field.replace(u=field.u * jnp.exp(1j * phase)) # showing that replace works 

@jax.jit
def forward(u):
    field = CoherentScalarField(0.25, 0.532, u) 
    return phase_change(field, 1.0)

# This all works!
field = jax.vmap(forward)(jnp.ones((5, 512, 512)))
print(field.intensity.shape) # [5, 512, 512]
print(field.power.shape) # [5, ]
print(field.grid.shape) # [5, 512, 512, 2]
print(field.k_grid.shape) # [5, 512, 512, 2]

# But this too!
@jax.jit
def forward(u):
    field = SpectralCoherentScalarField(0.25, [0.1, 0.532, 1.0], [0.2, 0.4, 0.1], u) 
    return phase_change(field, 1.0)

field = jax.vmap(forward)(jnp.ones((5, 512, 512, 3)))
print(field.intensity.shape) #[5, 512, 512, 3]
print(field.power.shape) #[5, 3], power / wavelength
print(field.grid.shape) #[5, 512, 512, 1, 2] # 1 because all dx are the same
print(field.k_grid.shape) #[5, 512, 512, 1, 2]
```

The full code and testcases can be found in the `dev` folder. I still need to add more test cases, but it seems this approach to do everything we'd like, fix all our issues, simplify code and do so without breaking changes. I'll add more stuff slowly, but would be happy to get some feedback here, @diptodip @ebezzam!

### CODE REVIEW

**Comment by diptodip** [comment]
This is great and I'll take a closer look soon, but I would probably advertise this as a large breaking change :)

File: pyproject.toml [hunk 1]
```diff
@@ -7,21 +7,34 @@ name = "chromatix"
...
+dependencies = [
+    "jax >= 0.6.0",
+    "einops >= 0.6.0",
+    "flax >= 0.6.3",
+    "chex>=0.1.5",
+    "optax >=0.1.4",
+    "scipy >= 1.10.0",
+    "equinox>=0.11.12",
+]
 version = "0.3.0"
```
Comment (ebezzam) [inline pyproject.toml:21]: Update version number for new Field object?

**Review by ebezzam — CHANGES_REQUESTED** [review]
@GJBoth thanks for this big switch! It does seem like a cleaner interface for making different types of field objects (Scalar or Vector, Monochromatic or Polychromatic). And the Spectrum object is a nice addition.

I installed the new dependencies and ran `test.py` and seems to work as expected. Last try-except are meant to fail right?

For now, my comments are mainly on refactoring and interface suggestions/ideas.

Would you like me to add bandlimited propagation to try working with the new Field object?

I agree with @diptodip that this should be advertised as a large breaking change. And I'm wondering if it would be good to push versions of chromatix to PyPI? I see that Dip did one a while back: https://pypi.org/project/chromatix/

It might be good to have a version with the old Field object as a PyPI package, instead of installing from GitHub, as in [current instructions](https://chromatix.readthedocs.io/en/latest/installing/#using-pip). There is a way to install from a specific git commit, but I think it would be cleaner to install from PyPI if people want to work with an older version of the Field object.
